src/sentry/snuba/query_subscription_consumer.py

```diff
@@ -0,0 +1,166 @@
+    return inner
+
+
+class InvalidMessageError(Exception):
+    pass
+
+
+class InvalidSchemaError(InvalidMessageError):
+    pass
+
+
+class QuerySubscriptionConsumer(object):
+    """
+    A Kafka consumer that processes query subscription update messages. Each message has
+    a related subscription id and the latest values related to the subscribed query.
+    These values are passed along to a callback associated with the subscription.
+    """
+
+    def __init__(
+        self,
+        group_id,
+        topic=None,
+        commit_batch_size=100,
+        initial_offset_reset='earliest',
+    ):
+        self.group_id = group_id
+        if not topic:
+            topic = settings.KAFKA_SNUBA_QUERY_SUBSCRIPTIONS
+        self.topic = topic
+        cluster_name = settings.KAFKA_TOPICS[topic]['cluster']
+        self.bootstrap_servers = settings.KAFKA_CLUSTERS[cluster_name]['bootstrap.servers']
+        self.commit_batch_size = commit_batch_size
+        self.initial_offset_reset = initial_offset_reset
+
+    def run(self):
+        logger.debug('Starting snuba query subscriber')
+
+        conf = {
+            'bootstrap.servers': self.bootstrap_servers,
+            'group.id': self.group_id,
+            'session.timeout.ms': 6000,
+            'auto.offset.reset': self.initial_offset_reset,
+            'enable.auto.commit': 'false',
+            'enable.auto.offset.store': 'true',
+            'enable.partition.eof': 'false',
+            'default.topic.config': {'auto.offset.reset': self.initial_offset_reset},
+        }
+        consumer = Consumer(conf)
+        consumer.subscribe([self.topic])
+        try:
+            i = 0
+            while True:
+                message = consumer.poll(0.1)
+                if message is None:
+                    continue
+
+                error = message.error()
+                if error is not None:
+                    raise KafkaException(error)
+
+                i = i + 1
+
+                self.handle_message(message)
+
+                if i % self.commit_batch_size == 0:
+                    logger.debug('Committing offsets')
+                    consumer.commit()
+        except KeyboardInterrupt:
+            pass
+        logger.debug('Committing offsets and closing consumer')
+        consumer.commit()
+        consumer.close()
+
+    def handle_message(self, message):
+        """
+        Parses the value from Kafka, and if valid passes the payload to the callback defined by the
+        subscription. If the subscription has been removed, or no longer has a valid callback then
+        just log metrics/errors and continue.
+        :param message:
+        :return:
+        """
+        try:
+            contents = self.parse_message_value(message.value())
+        except InvalidMessageError:
+            # If the message is in an invalid format, just log the error
+            # and continue
+            logger.exception('Subscription update could not be parsed')
```

Comment:
You'll probably want to log the partition and offset at least so you can go back and examine (or replay) the message. They aren't in the `Message.__repr__`, unfortunately (and annoyingly.)